### PR TITLE
Add invalid rule primary option message for `false`

### LINKED ISSUE #5091: Fix invalid option message for false for better DX
> Clearly describe the bug

Extending the standard config StyleLint gives me an error about the unit after a zero value. This cannot be undone disabling the corresponding rule because that causes another error.

> Which rule, if any, is the bug related to?

`length-zero-no-unit`

> What code is needed to reproduce the bug?

```css
.foobar {
  margin: 0px; /* Unexpected unit (length-zero-no-unit) */
}
```
> What stylelint configuration is needed to reproduce the bug?

```js
module.exports = {
  extends: "stylelint-config-standard",
  rules: {
    "length-zero-no-unit": false
  }
};
```

> Which version of stylelint are you using?

13.7.2

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

CLI with `stylelint test.css --config stylelint.config.js`

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

No.

> What did you expect to happen?

Being able to disable the rule and not get the complaint about the unit after a zero value, anymore.

> What actually happened (e.g. what warnings or errors did you get)?

```plain
Unexpected option value "false" for rule "length-zero-no-unit"
```

### PR DESCRIPTION
**Pull request by Mouvedia**
> Which issue, if any, is this issue related to?

Closes #5091
 
see also #2027 #5221 #3516 #1953

> Is there anything in the PR that needs further explanation?

~check the message of the first commit~
https://github.com/stylelint/stylelint/commit/101f2740f85315cb48f27714cfe4119b79beb6a3#diff-ea40f7b7d9827960e2d8e1590b37c7490648acb555d8fe8c22d0c2328449c95fR64

### CODE REVIEW

**Comment by Mouvedia** [comment]
This line
`for (const a of [actual].flat()) {`
in validateOptions.js motivated the change to the config.
Don't squash the commits so that it can be reverted easily.

**Comment by Mouvedia** [comment]
WIP
gotta fix the tests

**Comment by ybiquitous** [comment]
@Mouvedia Thanks for creating the PR!

The 66cdca0f709848f7e11a1e461b66ae07822766ba commit looks good to me. 👍🏼 

> Don't squash the commits so that it can be reverted easily.

We usually "squash and merge", so if you open a new PR including the commit above, I'll approve and merge it soon.

File: lib/utils/validateOptions.js [hunk 1]
```diff
@@ -65,6 +65,17 @@ function validate(opts, ruleName, complain) {
...
 	const optional = opts.optional;
+	const whitelist = [
+		'reportDescriptionlessDisables',
+		'reportInvalidScopeDisables',
+		'reportNeedlessDisables',
+	];
+
+	if (actual === false && !whitelist.includes(ruleName)) {
+		return complain(
+			`Invalid option value \`false\` for rule "${ruleName}". Are you trying to disable this rule? If so use \`null\` instead`,
```
Comment (ybiquitous) [inline lib/utils/validateOptions.js:76]: [suggest] `validateOptions` uses `"` consistently to embed any values into a string,

https://github.com/stylelint/stylelint/blob/0bd6976fea7fa2a07c47672de424194716d31164/lib/utils/validateOptions.js#L180-L186

so I think using `"` is better for `false` or `null`.

```suggestion
			`Invalid option value "false" for rule "${ruleName}". Are you trying to disable this rule? If so use "null" instead`,
```

File: lib/utils/validateOptions.js [hunk 2]
```diff
@@ -65,6 +65,17 @@ function validate(opts, ruleName, complain) {
 	const possible = opts.possible;
 	const actual = opts.actual;
 	const optional = opts.optional;
+	const whitelist = [
+		'reportDescriptionlessDisables',
+		'reportInvalidScopeDisables',
+		'reportNeedlessDisables',
+	];
```
Comment (ybiquitous) [inline lib/utils/validateOptions.js:72]: I think it's reasonable that `validateDisableSettings` uses internally `validateOptions` to re-use the logic of `validateOptions`:

https://github.com/stylelint/stylelint/blob/0bd6976fea7fa2a07c47672de424194716d31164/lib/validateDisableSettings.js#L50-L63

However, I don't think it's maintainable the `whitelist` array depends implicitly on the callees of `validateDisableSettings`; we would have to change two locations if we would add a new disable setting: 

https://github.com/stylelint/stylelint/blob/0bd6976fea7fa2a07c47672de424194716d31164/lib/invalidScopeDisables.js#L13

https://github.com/stylelint/stylelint/blob/0bd6976fea7fa2a07c47672de424194716d31164/lib/needlessDisables.js#L17

https://github.com/stylelint/stylelint/blob/0bd6976fea7fa2a07c47672de424194716d31164/lib/descriptionlessDisables.js#L16-L19

Is there a better way?

Comment (Mouvedia) [reply lib/utils/validateOptions.js:72]: > Is there a better way?

If there's a way to know the direct parent, yes there is.
In short, create a function `isRule`.

**Comment by Mouvedia** [comment]
@ybiquitous is
https://github.com/stylelint/stylelint/blob/060310cfd6383175ee06fe4d15aee27a4e202671/lib/__tests__/fixtures/plugin-async.js#L10
a problem?

**Comment by ybiquitous** [comment]
Ah, `possible: [true, null]` is less confusing. Can you fix it also, please?

File: lib/utils/validateOptions.js (hunk 2 above)
Comment (ybiquitous) [reply lib/utils/validateOptions.js]: Nice idea 👍🏼

File: lib/utils/validateOptions.js [hunk 3]
```diff
@@ -66,6 +66,12 @@ function validate(opts, ruleName, complain) {
 	const actual = opts.actual;
 	const optional = opts.optional;
 
+	if (actual === false && !ruleName.startsWith('report')) {
```
Comment (ybiquitous) [inline lib/utils/validateOptions.js:69]: [suggest] General rule names usually contain at least one hyphen (`-`),

https://github.com/stylelint/stylelint/blob/060310cfd6383175ee06fe4d15aee27a4e202671/docs/user-guide/rules/about.md?plain=1#L54-L57

so how about also checking if the characters are included for accuracy?

```suggestion
    // Ignoring options like `reportNeedlessDisables`
	if (actual === false && !ruleName.startsWith('report') && ruleName.includes("-")) {
```

(IMO, an auxiliary comment will help us understand this code.)

**Comment by Mouvedia** [comment]
> Ah, `possible: [true, null]` is less confusing. Can you fix it also, please?

Nop: `Type 'boolean' is not assignable to type 'RuleOptionsPossible[]'.`

File: lib/utils/validateOptions.js (hunk 3 above)
Comment (Mouvedia) [reply lib/utils/validateOptions.js:69]: > `&& ruleName.includes("-")`

That's wont work for `indentation` and `linebreaks`.

File: lib/rules/selector-no-qualifying-type/index.js [hunk 4]
```diff
@@ -63,7 +63,7 @@ const rule = (primary, secondaryOptions) => {
 			ruleName,
 			{
 				actual: primary,
-				possible: [true, false],
+				possible: [true],
```
Comment (ybiquitous) [inline lib/rules/selector-no-qualifying-type/index.js:66]: Good catch! This is a very minor bug. 😅

**Review by ybiquitous — APPROVED** [review]
Thank you. LGTM! 👍🏼

**Comment by ybiquitous** [comment]
Changelog entry added:

> - Added: invalid rule primary option message for `false` ([#6250](https://github.com/stylelint/stylelint/pull/6250)).

**Pull request merged**